h/security/permits/hybrid.py

```diff
@@ -0,0 +1,144 @@
+from enum import Enum
+
+from h.security.permits.predicates import *
+
+
+class PermissionEnum(Enum):
+    def __init__(self, _value, predicates):
+        self.predicates = predicates
+
+
+class Group(PermissionEnum):
+    ADMIN = (
+        "group:admin",
+        [
+            [group_authority_matches_authenticated_client],
+            [group_created_by_user],
+            # Those with the admin or staff Role should be able to admin/edit any
+            # group
+            [user_is_staff],
+            [user_is_admin],
+        ],
+    )
+    JOIN = (
+        "group:join",
+        [[group_joinable_by_authority, group_matches_user_authority]],
+    )
+    READ = (
+        "group:read",
+        [
+            [group_readable_by_world],
+            [group_readable_by_members, group_has_user_as_member],
+            [group_authority_matches_authenticated_client],
+        ],
+    )
+    MEMBER_READ = (
+        "group:member:read",
+        [
+            [group_readable_by_world],
+            [group_readable_by_members, group_has_user_as_member],
+            [group_authority_matches_authenticated_client],
+        ],
+    )
+    WRITE = (
+        "group:write",
+        [
+            [group_writable_by_authority, group_matches_user_authority],
+            [group_writable_by_members, group_has_user_as_member],
+        ],
+    )
+    UPSERT = (
+        "group:upsert",
+        [
+            [group_created_by_user],
+            [group_not_found, authenticated_user],
+        ],
+    )
+    CREATE = ("group:create", [[authenticated_user]])
+    FLAG = (
+        "group:flag",
+        [
+            # Any logged in user should be able to flag things they can see
+            [group_readable_by_world, authenticated],
+            [group_readable_by_members, group_has_user_as_member],
+        ],
+    )
+    MODERATE = ("group:moderate", [[group_created_by_user]])
+    MEMBER_ADD = (
+        "group:member:add",
+        [[group_authority_matches_authenticated_client]],
+    )
+
+
+class Annotation(PermissionEnum):
+    CREATE = ("annotation:create", [[authenticated]])
+    READ = (
+        "annotation:read",
+        [
+            [annotation_live, annotation_not_shared, annotation_created_by_user],
+            [annotation_live, annotation_shared, Group.READ],
```

Comment:
This annoyance of refering back to the `Group` permissions forces us to split up the code a bit. We can't do it in the old inner class style, because the `Group` class isn't defined yet when we do it that way. Suggestions welcome.